### PR TITLE
Fix incorrect multiline indentation inserting non-snippet text

### PR DESCRIPTION
**Pull request by zardoy**
I added failing tests to illustrate the problem.

I tried different approaches, and it seems only `insertText` of completion and `editor.insertText` does whitespace adjusting in VSCode. First one causes [editor scroll jump higher](https://github.com/ipatalas/vscode-postfix-ts/pull/73), and a second one messes up with undo stack. So I tried to do whitespace normalization manually, however:

1. It doesn't seem to be logical that we first remove all whitespace characters (`adjustMultilineIndentation`) and then add them back (code that adds this pr).
2. There is an issue with `.not`:

```ts
    if (
        a && (b &&
        a
        .a
        .b).not
    )  {}
```

### CODE REVIEW

File: test/extension.multiline.test.ts [hunk 1]
```diff
@@ -48,6 +48,32 @@ describe('Multiline template tests', () => {
       | \t.b        >> \t.b
       | \t.c++{let} >> \t.c++`)
 
+  Test(`return template - method call (equal indentation)
```
Comment (zardoy) [inline test/extension.multiline.test.ts:72]: So I just duplicated tests from `let` template above, as `return` snippet was inserting as non-snippet text.

File: src/completionItemBuilder.ts [hunk 2]
```diff
@@ -95,12 +97,16 @@ export class CompletionItemBuilder {
...
 
-    return replacement.replace(re, (_match, p1) => {
+    let replacedText = replacement.replace(re, (_match, p1) => {
       if (p1 && this.filters[p1]) {
         return this.filters[p1](code)
       }
       return code;
     })
+    const { line } = this.node.getSourceFile().getLineAndCharacterOfPosition(this.node.getStart())
+    const indentStart = vsc.window.activeTextEditor.document.lineAt(line).text.match(/^\s*/)?.[0]
```
Comment (zardoy) [inline src/completionItemBuilder.ts]: Yyea I forgot to refactor this, just wanted to check it works & didn't want to add additional arg to all templates.

Comment (zardoy) [reply src/completionItemBuilder.ts]: I think it should be passed from [getIndentSize](https://github.com/ipatalas/vscode-postfix-ts/blob/develop/src/postfixCompletionProvider.ts#L122) somehow.
By the way its always safe to put `\n` into snippet as editor always does the normalization.

**Comment by zardoy** [comment]
I also checked that my proposed workaround for snippets in https://github.com/ipatalas/vscode-postfix-ts/pull/73#issuecomment-1249724849 doesn't do whitespace adjustment as well.

Just leaving a reference: https://github.com/microsoft/vscode/blob/058e70bff799899843e110b8d263788dc3aa83be/src/vs/editor/contrib/snippet/browser/snippetSession.ts#L388 it seems to be straightforward

**Comment by zardoy** [comment]
@ipatalas Sorry I was completely killed last evening and as I see didn't describe things clearly here.

So with https://github.com/ipatalas/vscode-postfix-ts/pull/73 I broke inserting text for non-snippet templates (like console or return) when node had leading whitespace characters. As I described in the body, there is no reliable way to reuse this logic from vscode, but we can do it manually since its not complicated.

I finally managed to provide my final solution to this. The only issue is with `.not` template:

tl;dr I do not understand why https://github.com/ipatalas/vscode-postfix-ts/blob/eb7f989d7c3b9f5d2641f75007a29367cfd3ff98/src/utils/invert-expression.ts#L36
this is being called on text that is returned only in some cases. Why we don't normalize final text result?

Also let me know what you think about it, as it touches every template.

**Comment by ipatalas** [comment]
Hi, sorry, I already had a very simple solution for that few hours ago but had to go offline :)
Anyway I think that fixes the problem: https://github.com/ipatalas/vscode-postfix-ts/commit/c66c8f6213244cd4b067bc8bb6f7b5f6ed48037f
All tests are green and it's one liner fix :D The thing is that I was already trying to fix whitespace problems with that `adjustMultilineIndentation` function. 
For snippet strings vscode does that on its own as you mentioned so now I just turned off this `adjustMultilineIndentation` for non-snippet completions and apparently it works quite well.

It has small issue with your `.not` example but this might be related to what you discovered in that `invert-expression.ts`. I don't remember why it's not fixing the output. Perhaps at the beginning it was working fine, the code there was less complex and then it grew but this stayed because nobody complained ;)

**Comment by ipatalas** [comment]
Kind of fixed `.not` partially:
bug-83-not

The other alternative doesn't work perfectly (adds one extra new line which shouldn't be there) but maybe I'll find a way. Actually this fix means removing some code - I like that :D Perhaps VS Code API has changed many things since this `adjustMultilineIndentation` was first introduced. It looks like it's no longer needed in most places.

**Comment by zardoy** [comment]
> c66c8f6

Yeah, you might noticed that this PR has a bit more code changes, because I'm replicating VSCode's normalization behavior (that's why I'm wrapping output result, not `code`). Sorry if I overengineered this, didn't see 1 line solution here :)

> All tests are green and it's one liner fix

Sorry, but there are still cases that it doesn't fix:

```json
{
    "name": "custom",
    "body": "1\n\t1\n{{expr}}",
    "when": [
        "identifier",
        "expression",
        "function-call"
    ]
},
```
Or even better, without `{{expr}}`: `1\n\t1\n1`

Examples:

```
    a
    .b
    .c{custom}
```

Correct (just add `$0` to the end):

```
    1
		1
	a
	.b
	.c
```

Incorrect:

```
	1
	1
a
	.b
	.c
```

**Comment by ipatalas** [comment]
It's not a final solution, i didn't push last commit because I'm still trying to figure this out.
Anyway I'm advocating this solution to avoid having two places that adjust whitespaces.
I'd rather remove the old solution (which I sort of did) and then start over with the fixes.
Otherwise it will be hard to maintain if two functions are doing more or less the same in two different places.

**Comment by ipatalas** [comment]
All right, fixed that first problem:
bug-83-not-v2

The template is clearly reversible for both alternatives with no side effects.
It still won't work with your custom template example, it's another story because the template replacement is multi-line and contains whitespaces. That will be next step where this PR's code might come handy. I reviewed that quickly and the idea itself looks pretty good, I had the same idea with `IndentInfo` object to pass additional information.

**Comment by ipatalas** [comment]
Ok, pushed the changes and now `.not` should work much better with those whitespaces.
Be careful when resolving conflicts as we both did changes in the same lines.
I'll try to review and test this PR tomorrow and then it would also be done for custom templates 👍🏻

File: src/postfixCompletionProvider.ts [hunk 3]
```diff
@@ -43,6 +43,7 @@ export class PostfixCompletionProvider implements vsc.CompletionItemProvider {
     }
 
     const indentSize = this.getIndentSize(document, currentNode)
+    const leadingWhitespace = this.getLeadingWhitespace(document, currentNode)
```
Comment (ipatalas) [inline src/postfixCompletionProvider.ts]: How about merging those two into `getIndentInfo`? They are very similar inside.

File: test/extension.multiline.test.ts [hunk 4]
```diff
@@ -48,6 +48,32 @@ describe('Multiline template tests', () => {
...
+      | ${indent(2)}object.call()   >> ${indent(2)}return object.call()
+      | ${indent(3)}.anotherCall()  >> \t\t\t.anotherCall()
+      | ${indent(3)}.lastOne(){return} >> \t\t\t.lastOne()`)
+
+  Test(`return template - method call (indentation - mixed)
+      | \t\tobject.call()          >> \t\treturn object.call()
+      | ${indent(3)}.anotherCall() >> \t\t\t.anotherCall()
+      | \t\t\t.lastOne(){return}      >> \t\t\t.lastOne()`)
+
+  Test(`return template - method call (indentation - compreturnely mixed)
```
Comment (ipatalas) [inline test/extension.multiline.test.ts]: `compreturnely`? :)

File: src/postfixCompletionProvider.ts (hunk 3 above)
Comment (zardoy) [reply src/postfixCompletionProvider.ts]: To be honest I didn't even notice this somehow. Thank you for suggestion!

File: test/extension.multiline.test.ts (hunk 4 above)
Comment (zardoy) [reply test/extension.multiline.test.ts]: :D good catch, I was obviously using multicursor editing on these tests, though have no idea how I did it with the name

**Comment by zardoy** [comment]
Sorry for the delay, I somehow missed the fact that pr with `.not` template was merged.
I also added a multiline test with custom template that I mentioned above. Let me know what you think.

File: src/postfixCompletionProvider.ts [hunk 5]
```diff
@@ -105,28 +104,26 @@ export class PostfixCompletionProvider implements vsc.CompletionItemProvider {
     return currentNode
   }
 
-  private getIndentSize(document: vsc.TextDocument, node: ts.Node): number | undefined {
+  private getIndentInfo(document: vsc.TextDocument, node: ts.Node): IndentInfo | undefined {
```
Comment (ipatalas) [inline src/postfixCompletionProvider.ts]: Can this ever return `undefined`? I think it was like that for old `getIndentSize` but not this version anymore.
However maybe it's good to have a default value for `indentSize` because it won't be initialized when none of the conditions in lines 115-117 are satisfied.

Comment (zardoy) [reply src/postfixCompletionProvider.ts]: Of course it all make sense. Should be better now.

File: test/extension.multiline.test.ts [hunk 6]
```diff
@@ -37,25 +37,25 @@ describe('Multiline template tests', () => {
...
+      | ${indent(3)}.anotherCall()    >> \t\t\t.anotherCall()
       | \t\t\t.lastOne(){return}      >> \t\t\t.lastOne()`)
 
   Test(`return template - method call (indentation - completely mixed)
-      | \tobject.call()     >> \treturn object.call()
-      | \t  .anotherCall()  >> \t  .anotherCall()
-      | \t  .lastOne(){return} >> \t  .lastOne()`)
+      | \tobject.call()        >> \treturn object.call()
+      | \t  .anotherCall()     >> \t\t  .anotherCall()
+      | \t  .lastOne(){return} >> \t\t  .lastOne()`)
```
Comment (ipatalas) [inline test/extension.multiline.test.ts:58]: Sorry, missed that last time but now noticed you added additional tabs. Original indent was one tab only so why two here? Shouldn't that still be one tab + spaces?

Comment (zardoy) [reply test/extension.multiline.test.ts:58]: I just copied it from [let template](https://github.com/ipatalas/vscode-postfix-ts/pull/83/files/b04272e23b5d595e95b37cfbc7d396ee8b49149c#diff-c04594fbca6775313bf1259faca5b5bea36bd1cd98ef510b6458158ec292436dL34-L37) the only difference is that I replaced it with `return`

**Review by ipatalas — APPROVED** [review]

**Comment by zardoy** [comment]
@ipatalas Hey! It'd be great if you published this fix on marketplace :)

By the way, I'm just curious do you personally use this extension, or you moved away from TS development ?

**Comment by ipatalas** [comment]
> @ipatalas Hey! It'd be great if you published this fix on marketplace :)

Hey, sorry, somehow thought I did that already. Anyway, just published.

> By the way, I'm just curious do you personally use this extension, or you moved away from TS development ?

Unfortunately not at the moment. 
I used to be a fullstack and then I was using it extensively in day to day work but currently I do almost only BE at my current project hence I don't have a need to use it. 
Of course I have it installed and ready to be used whenever I have anything JS/TS related to do but it happens only once or twice a month.
